README.md
<img src="https://raw.githubusercontent.com/sergioburdisso/pyss3/master/docs/_static/ss3_logo_banner.png" alt="PySS3 Logo" title="PySS3" height="150" />

[![Documentation Status](https://readthedocs.org/projects/pyss3/badge/?version=latest)](http://pyss3.readthedocs.io/en/latest/?badge=latest)
[![Build Status](https://api.travis-ci.com/sergioburdisso/pyss3.svg?branch=master)](https://app.travis-ci.com/github/sergioburdisso/pyss3)
[![codecov](https://codecov.io/gh/sergioburdisso/pyss3/branch/master/graph/badge.svg)](https://codecov.io/gh/sergioburdisso/pyss3)
[![Requirements Status](https://requires.io/github/sergioburdisso/pyss3/requirements.svg?branch=master)](https://requires.io/github/sergioburdisso/pyss3/requirements/?branch=master)
[![PyPI version](https://badge.fury.io/py/pyss3.svg)](https://badge.fury.io/py/pyss3)
[![Downloads](https://pepy.tech/badge/pyss3)](https://pepy.tech/project/pyss3)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/sergioburdisso/pyss3/master?filepath=examples)

---

# A Python package implementing a new simple and interpretable model for text classification

:sushi: **Online live demos:** http://tworld.io/ss3/ :icecream::ice_cream::cake:

<br>

The SS3 text classifier is a novel and simple supervised machine learning model for text classification which is interpretable, that is, it has the **ability to naturally (self)explain its rationale**. It was originally introduced in Section 3 of the paper _["A text classification framework for simple and effective early depression detection over social media streams"](https://dx.doi.org/10.1016/j.eswa.2019.05.023)_ ([arXiv preprint](https://arxiv.org/abs/1905.08772)).
This simple model obtained the best and 2nd-best results, consecutively, in the last three editions of the [CLEF's eRisk lab](https://erisk.irlab.org/) among all participating models [[Burdisso *et al.* 2019](http://ceur-ws.org/Vol-2380/paper_103.pdf); [Loyola *et al.* 2021](http://ceur-ws.org/Vol-2936/paper-81.pdf)].
Given its white-box nature, it allows researchers and practitioners to deploy interpretable (i.e. self-explainable) and therefore more reliable, models for text classification (which could be especially useful for those working with classification problems by which people's lives could be somehow affected).

**Note:** this package also incorporates different variations of the original model, such as the one introduced in _["t-SS3: a text classifier with dynamic n-grams for early risk detection over text streams"](https://doi.org/10.1016/j.patrec.2020.07.001)_ ([arXiv preprint](https://arxiv.org/abs/1911.06147)) which allows SS3 to recognize important variable-length word n-grams "on the fly".

## What is PySS3?

[PySS3](https://github.com/sergioburdisso/pyss3) is a Python package that allows you to work with SS3 in a very straightforward, interactive and visual way. In addition to the implementation of the SS3 classifier, PySS3 comes with a set of tools to help you developing your machine learning models in a clearer and faster way. These tools let you analyze, monitor and understand your models by allowing you to see what they have actually learned and why. To achieve this, PySS3 provides you with 3  main components: the ``SS3`` class, the ``Live_Test`` class, and the ``Evaluation`` class, as pointed out below.


### :point_right: The ``SS3`` class

which implements the classifier using a clear API (very similar to that of `sklearn`'s models):
````python
    from pyss3 import SS3
    clf = SS3()
    ...
    clf.fit(x_train, y_train)
    y_pred = clf.predict(x_test)
````

Also, this class provides a handful of other useful methods, such as, for instance, [``extract_insight()``](https://pyss3.rtfd.io/en/latest/api/index.html#pyss3.SS3.extract_insight) to [extract the text fragments involved in the classification decision](https://pyss3.readthedocs.io/en/latest/tutorials/extract-insight.html) (allowing you to better understand the rationale behind the model’s predictions) or [``classify_multilabel()``](https://pyss3.rtfd.io/en/latest/api/index.html#pyss3.SS3.classify_multilabel) to provide [multi-label classification](https://en.wikipedia.org/wiki/Multi-label_classification) support: 

````python
    doc = "Liverpool CEO Peter Moore on Building a Global Fanbase"
    
    # standard "single-label" classification
    label = clf.classify_label(doc) # 'business'

    # multi-label classification
    labels = clf.classify_multilabel(doc)  # ['business', 'sports']
````

### :point_right: The ``Live_Test`` class

which allows you to interactively test your model and visually see the reasons behind classification decisions, **with just one line of code**:
```python
    from pyss3.server import Live_Test
    from pyss3 import SS3

    clf = SS3()
    ...
    clf.fit(x_train, y_train)
    Live_Test.run(clf, x_test, y_test) # <- this one! cool uh? :)
```
As shown in the image below, this will open up, locally, an interactive tool in your browser which you can use to (live) test your models with the documents given in `x_test` (or typing in your own!). This will allow you to visualize and understand what your model is actually learning.

![img](https://raw.githubusercontent.com/sergioburdisso/pyss3/master/docs/_static/ss3_live_test.gif)

For example, we have uploaded two of these live tests online for you to try out: ["Movie Review (Sentiment Analysis)"](http://tworld.io/ss3/live_test_online/#30305) and ["Topic Categorization"](http://tworld.io/ss3/live_test_online/#30303), both were obtained following the [tutorials](https://pyss3.readthedocs.io/en/latest/user_guide/getting-started.html#tutorials).

### :point_right: And last but not least, the ``Evaluation`` class

This is probably one of the most useful components of PySS3. As the name may suggest, this class provides the user easy-to-use methods for model evaluation and hyperparameter optimization, like, for example, the [``test``](https://pyss3.readthedocs.io/en/latest/api/index.html#pyss3.util.Evaluation.test), [``kfold_cross_validation``](https://pyss3.readthedocs.io/en/latest/api/index.html#pyss3.util.Evaluation.kfold_cross_validation), [``grid_search``](https://pyss3.readthedocs.io/en/latest/api/index.html#pyss3.util.Evaluation.grid_search), and [``plot``](https://pyss3.readthedocs.io/en/latest/api/index.html#pyss3.util.Evaluation.plot) methods for performing tests, stratified k-fold cross validations, grid searches for hyperparameter optimization, and visualizing evaluation results using an interactive 3D plot, respectively. Probably one of its most important features is the ability to automatically (and permanently) record the history of evaluations that you've performed. This will save you a lot of time and will allow you to interactively visualize and analyze your classifier performance in terms of its different hyper-parameters values (and select the best model according to your needs). For instance, let's perform a grid search with a 4-fold cross-validation on the three [hyperparameters](https://pyss3.rtfd.io/en/latest/user_guide/ss3-classifier.html#hyperparameters), smoothness(`s`), significance(`l`), and sanction(`p`):

```python
from pyss3.util import Evaluation
...
best_s, best_l, best_p, _ = Evaluation.grid_search(
    clf, x_train, y_train,
    s=[0.2, 0.32, 0.44, 0.56, 0.68, 0.8],
    l=[0.1, 0.48, 0.86, 1.24, 1.62, 2],
    p=[0.5, 0.8, 1.1, 1.4, 1.7, 2],
    k_fold=4
)
```
In this illustrative example, `s`, `l`, and `p` will take those 6 different values each, and once the search is over, this function will return (by default) the hyperparameter values that obtained the best accuracy.
Now, we could also use the ``plot`` function to analyze the results obtained in our grid search using the interactive 3D evaluation plot:

```python
Evaluation.plot()
```

![img](https://raw.githubusercontent.com/sergioburdisso/pyss3/master/docs/_static/plot_evaluations.gif)

In this 3D plot, each point represents an experiment/evaluation performed using that particular combination of values (s, l, and p). Also, these points are painted proportional to how good the performance was according to the selected metric; the plot will update "on the fly" when the user select a different evaluation metric (accuracy, precision, recall, f1, etc.). Additionally, when the cursor is moved over a data point, useful information is shown (including a "compact" representation of the confusion matrix obtained in that experiment). Finally, it is worth mentioning that, before showing the 3D plots, PySS3 creates a single and portable HTML file in your project folder containing the interactive plots. This allows users to store, send or upload the plots to another place using this single HTML file. For example, we have uploaded two of these files for you to see: ["Sentiment Analysis (Movie Reviews)"](https://pyss3.readthedocs.io/en/latest/_static/ss3_model_evaluation[movie_review_3grams].html) and ["Topic Categorization"](https://pyss3.readthedocs.io/en/latest/_static/ss3_model_evaluation[topic_categorization_3grams].html), both evaluation plots were also obtained following the [tutorials](https://pyss3.readthedocs.io/en/latest/user_guide/getting-started.html#tutorials).


## Want to give PySS3 a shot? :eyeglasses: :coffee:

Just go to the [Getting Started](https://pyss3.readthedocs.io/en/latest/user_guide/getting-started.html) page :D

### Installation


Simply use:
```console
pip install pyss3
```

## Want to contribute to this Open Source project? :sparkles::octocat::sparkles:

Thanks for your interest in the project, you're ![Awesome](https://cdn.rawgit.com/sindresorhus/awesome/d7305f38d29fed78fa85652e3a63e154dd8e8829/media/badge.svg)!!
Any kind of help is very welcome (Code, Bug reports, Content, Data, Documentation, Design, Examples, Ideas, Feedback, etc.),  Issues and/or Pull Requests are welcome for any level of improvement, from a small typo to new features, help us make PySS3 better :+1:

Remember that you can use the "Edit" button ('pencil' icon) up the top to [edit any file of this repo directly on GitHub](https://help.github.com/en/github/managing-files-in-a-repository/editing-files-in-your-repository).

Finally, in case you're planning to create a **new Pull Request**, for committing to this repo, we follow the "seven rules of a great Git commit message" from ["How to Write a Git Commit Message"](https://chris.beams.io/posts/git-commit/), so make sure your commits follow them as well.

(if you need any further information, please, **do not hesitate** to contact me - sergio.burdisso@gmail.com)

### Contributors :muscle::sunglasses::+1:

Thanks goes to these awesome people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="http://angermeir.me/"><img src="https://avatars3.githubusercontent.com/u/16398152?v=4" width="100px;" alt=""/><br /><sub><b>Florian Angermeir</b></sub></a><br /><a href="#code-angrymeir" title="Code">💻</a> <a href="#ideas-angrymeir" title="Ideas, Suggestions, & Feedback">🤔</a> <a href="#data-angrymeir" title="Data">🔣</a></td>
    <td align="center"><a href="https://www.linkedin.com/in/muneebvaiyani/"><img src="https://avatars3.githubusercontent.com/u/36028992?v=4" width="100px;" alt=""/><br /><sub><b>Muneeb Vaiyani</b></sub></a><br /><a href="#ideas-Vaiyani" title="Ideas, Suggestions, & Feedback">🤔</a> <a href="#data-Vaiyani" title="Data">🔣</a></td>
    <td align="center"><a href="https://www.saurabhbora.com"><img src="https://avatars2.githubusercontent.com/u/29205181?v=4" width="100px;" alt=""/><br /><sub><b>Saurabh Bora</b></sub></a><br /><a href="#ideas-enthussb" title="Ideas & Feedback">🤔</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Further Readings :scroll:


[Full documentation](https://pyss3.readthedocs.io)

[API documentation](https://pyss3.readthedocs.io/en/latest/api/)

[Paper preprint](https://arxiv.org/abs/1912.09322)


CHANGELOG.md
# ChangeLog

All notable changes to PySS3 will be documented here.

## [0.6.4] 2021-01-30

### Fixed

- Quick fix of default compatibility with foreign languages ([#15](https://github.com/sergioburdisso/pyss3/issues/15)).


## [0.6.3] 2020-07-17

### Fixed

- Patches issue [#11](https://github.com/sergioburdisso/pyss3/issues/11).

## [0.6.1] 2020-05-26

### Added
- ``Dataset.load_from_files_multilabel()`` can load documents with no
  labels as well ([31251f8](https://github.com/sergioburdisso/pyss3/commit/31251f8)).

- A ``set_testset_from_files_multilabel()`` function was added to the
  ``Live_Test`` class. This function allows loading multilabel
  datasets from disk Live Test server ([0ddbd6a](https://github.com/sergioburdisso/pyss3/commit/0ddbd6a)).

### Fixed

- Fixed a bug in SS3 hyperparameter initialization ([e2e72f9](https://github.com/sergioburdisso/pyss3/commit/e2e72f9)).

## [0.6.0] 2020-05-24

### Added

PySS3 now fully support multi-label classification! :)

- The ``load_from_files_multilabel()`` function was added to the ``Dataset`` class ([7ece7ce](https://github.com/sergioburdisso/pyss3/commit/7ece7ce), resolved [#6](https://github.com/sergioburdisso/pyss3/issues/6))

- The ``Evaluation`` class now supports multi-label classification (resolved [#5](https://github.com/sergioburdisso/pyss3/issues/5))
  - Add multi-label support to ``train()/fit()`` ([4d00476](https://github.com/sergioburdisso/pyss3/commit/4d00476))
  - Add multi-label support to ``Evaluation.test()`` ([0a897dd](https://github.com/sergioburdisso/pyss3/commit/0a897dd))
  - Add multi-label support to ``show_best and get_best()`` ([ef2419b](https://github.com/sergioburdisso/pyss3/commit/ef2419b))
  - Add multi-label support to ``kfold_cross_validation()`` ([aacd3a0](https://github.com/sergioburdisso/pyss3/commit/aacd3a0))
  - Add multi-label support to ``grid_search()`` ([925156d](https://github.com/sergioburdisso/pyss3/commit/925156d), [79f1e9d](https://github.com/sergioburdisso/pyss3/commit/79f1e9d))
  - Add multi-label support to the 3D Evaluation Plot ([42bbc65](https://github.com/sergioburdisso/pyss3/commit/42bbc65))

- The Live Test tool now supports multi-label classification as well ([15657ee](https://github.com/sergioburdisso/pyss3/commit/15657ee), [b617bb7](https://github.com/sergioburdisso/pyss3/commit/b617bb7), resolved [#9](https://github.com/sergioburdisso/pyss3/issues/9))

- Category names are no longer case-insensitive ([4ec009a](https://github.com/sergioburdisso/pyss3/commit/4ec009a), resolved [#8](https://github.com/sergioburdisso/pyss3/issues/8))

## [0.5.7] 2020-05-05

### Added

- The Live Test Tool now supports custom (user-defined) preprosessing methods ([b50cfaf](https://github.com/sergioburdisso/pyss3/commit/b50cfaf), [7c6b0c6](https://github.com/sergioburdisso/pyss3/commit/7c6b0c6), resolved [#3](https://github.com/sergioburdisso/pyss3/issues/3)).

- The tokenization process was improved ([26fff88](https://github.com/sergioburdisso/pyss3/commit/26fff88), [4af8e80](https://github.com/sergioburdisso/pyss3/commit/4af8e80)).

- The process for recognizing word n-grams during classification was improved ([2ceb148](https://github.com/sergioburdisso/pyss3/commit/2ceb148)).

## [0.5.5] 2020-03-02

### Added

- The ``predict`` method was optimized. Now it is 10x to 200x faster! This improvement also has a positive impact on other methods that use ``predict`` such as ``grid_search`` ([37202d8](https://github.com/sergioburdisso/pyss3/commit/37202d8)).
- A new `get_ngrams_length` method was added to ``SS3`` class. It can be used to get the length of longest learned n-gram ([b4f8827](https://github.com/sergioburdisso/pyss3/commit/b4f8827)).
- The Evaluation 3D Plot's GUI was improved ([1bb1e5a](https://github.com/sergioburdisso/pyss3/commit/1bb1e5a)).

### Fixed

- Some bugs and error were fixed ([bc5c4ed](https://github.com/sergioburdisso/pyss3/commit/bc5c4ed), [0d3d7e1](https://github.com/sergioburdisso/pyss3/commit/0d3d7e1), [86a0189](https://github.com/sergioburdisso/pyss3/commit/86a0189), [b0b3eaa](https://github.com/sergioburdisso/pyss3/commit/b0b3eaa), [5dbdc3a](https://github.com/sergioburdisso/pyss3/commit/5dbdc3a))


## [0.5.0] 2020-02-24

### Added
- A new ``Evaluation`` class to ``pyss3.util`` ([8feeef5](https://github.com/sergioburdisso/pyss3/commit/8feeef5a44ccc26e98f967fe470d0d0521d97f96)).
    - Now the user can import the ``Evaluation`` class to perform model evaluation and hyperparameter optimization. This class not only provide methods to evaluate models but also keeps all the advantages previously provided only through the Command Line tool, such as an evaluation cache that automatically keeps track of the evaluation history and the generation of the interactive 3D evaluation plot.
- ``set_name()`` to ``SS3`` ([5b1c355](https://github.com/sergioburdisso/pyss3/commit/5b1c355070ad66884f4360128cbf4f97d9b018b6)).
- ``train()`` to ``SS3`` as a user-friendly alias of ``fit()`` ([74cb540](https://github.com/sergioburdisso/pyss3/commit/74cb54067e10dfeecf0bb52a05d20d2e84b3b34c)).
- Print now supports nested verbosity regions ([78176ab](https://github.com/sergioburdisso/pyss3/commit/78176abb27f2b8a4e7233820ab93265f5c4ee5d5)).

### Fixed

- Compatibility of progress bars with Jupyter Notebooks ([7848b3e](https://github.com/sergioburdisso/pyss3/commit/7848b3e97d42dfb4121ddddbf3fcbae9e9e6736e), [8d163d9](https://github.com/sergioburdisso/pyss3/commit/8d163d9c1b6391fd32c0c5fc0b6d2190376d7f1f), [2029c37](https://github.com/sergioburdisso/pyss3/commit/2029c37af1e7739865402f4af194cd7fc122a2f8), [2a700d5](https://github.com/sergioburdisso/pyss3/commit/2a700d53c5d676c5bbba2cc21494f596d05fbfd2)).
- Bug in SS3.fit when given an empty document ([31eccbc](https://github.com/sergioburdisso/pyss3/commit/31eccbcb193efd3c8ebdacbae12615f54528c37e)).
- Non-string category labels support ([5b1c355](https://github.com/sergioburdisso/pyss3/commit/5b1c355070ad66884f4360128cbf4f97d9b018b6)).

examples/topic_categorization.py
# coding: utf-8
"""
Topic Categorization Tutorial Source Code (Classic Workflow).

(https://pyss3.readthedocs.io/en/latest/tutorials/topic-categorization.html#topic_classic-workflow)
---
"""

# Before we begin, let's import needed modules...
from pyss3 import SS3
from pyss3.util import Dataset
from pyss3.server import Live_Test

from sklearn.metrics import accuracy_score
from os import system

# ... and unzip the "movie_review.zip" dataset inside the `datasets` folder.
system('unzip -u datasets/topic.zip -d datasets/')

# Ok, now we are ready to begin. Let's create a new SS3 instance
clf = SS3()


# What are the default hyperparameter values? let's see
s, l, p, _ = clf.get_hyperparameters()

print("Smoothness(s):", s)
print("Significance(l):", l)
print("Sanction(p):", p)


# Ok, now let's load the training and the test set using the `load_from_files` function
# from `pyss3.util`. Since, in this dataset, there's a single file for each category,
# we will use the argument ``folder_label=False`` to tell PySS3 to use each file as a different
# category and each line inside of it as a different document:
x_train, y_train = Dataset.load_from_files("datasets/topic/train", folder_label=False)
x_test, y_test = Dataset.load_from_files("datasets/topic/test", folder_label=False)


# Let's train our model...
clf.fit(x_train, y_train)


# Note that we don't have to create any document-term matrix! we are using just the
# plain `x_train` documents :D cool uh? (SS3 creates a language model for each category
# and therefore it doesn't need to create any document-term matrices)
#
# Now that the model has been trained, let's test it using the documents in `x_test`
y_pred = clf.predict(x_test)


# Let's see how good our model performed
print("Accuracy:", accuracy_score(y_pred, y_test))


# Not bad using the default hyperparameters values... let's manually analyze
# what this model has actually learned by using the interactive "live test".
# Note that since we are not going to use the `x_test` for this live test(*) but instead
# the documents in "datasets/topic/live_test", we must use the `set_testset_from_files`
# method to tell the server to load documents from there instead.
#
# (*) try it if you want but since `x_test` contains (preprocessed) tweets,
# they don't look really good and clean.
#
# Live_Test.run(clf, x_test, y_test)  # <- this visualization doesn't look really clean and good so,
#                                     #     instead, we will use the documents in "live_test" folder:

Live_Test.set_testset_from_files("datasets/topic/live_test")

Live_Test.run(clf)

# Live test doesn't look bad, however, we will create a "more intelligent" version of this model,
# a version that can recognize variable-length word n-grams "on the fly". Thus, when calling the
# `fit` we will pass an extra argument `n_grams=3` to indicate we want SS3 to learn to recognize
# important words, bigrams, and 3-grams(*). Additionally, we will name our model
# "topic_categorization_3grams" so that we can save it and load it later from the
# `PySS3 Command Line` to perform the hyperparameter optimization to find better
# hyperparameters values.
#
# (*) If you're curious and want to know how this is actually done by SS3,
#     read the paper "t-SS3: a text classifier with dynamic n-grams for early risk detection over
#     text streams" (preprint available [here](https://arxiv.org/abs/1911.06147)).
clf = SS3(name="topic_categorization_3grams")

clf.fit(x_train, y_train, n_grams=3)  # <-- note the n_grams=3 argument here


# As mentioned above, we will save this trained model for later use
clf.save_model()


# Now let's see if the performance has improved...
y_pred = clf.predict(x_test)
print("Accuracy:", accuracy_score(y_pred, y_test))


# Yeah, the accuracy slightly improved but more importantly, we should now see that the model
# has learned "more intelligent patterns" involving sequences of words when using the
# interactive "live test" to observe what our model has learned (like "machine learning",
# "artificial intelligence", "self-driving cars", etc. for the "science&technology" category.
# Let's see...
Live_Test.run(clf)

# Fortunately, our model has learned to recognize these important sequences (such as
# "artificial intelligence" and "machine learning" in doc_2.txt,  "self-driving cars"
# in doc_6.txt, etc.). However, some documents aren’t perfectly classified, for instance,
# doc_3.txt was classified as “science&technology” (as a third topic) which is clearly wrong...
#
# So, one last thing we are going to do is to try yo find better hyperparameter values to improve
# our model's performance. To achieve this, we will perform what its known as
# "Hyperparameter Optimization" using the ``PySS3 Command Line`` tool.

# * At this point you should read the "Hyperparameter Optimization" section of this tutorial.
# (https://pyss3.readthedocs.io/en/latest/tutorials/topic-categorization.html.html#topic_hyperparameter-optimization)
#
# As described in the "Hyperparameter Optimization" section, we found out that the following
# hyperparameter values will improve our classification performance
clf.set_hyperparameters(s=0.32, l=1.24, p=1.1)


# Let's see if it's true...
y_pred = clf.predict(x_test)
print("Accuracy:", accuracy_score(y_pred, y_test))


# The accuracy has improved as expected :)
#
# Let's perform the last check and visualize what our final model has
# learned and how it is classifying the documents...
Live_Test.run(clf)


# Perfect! now the documents are classified properly! (including *doc_3.txt*) :D
#
# ...and that's it, nicely done buddy!
